be/log_discard.c

```diff
@@ -230,72 +227,73 @@ static void be_log_discard_item_discard_ast(struct m0_sm_group *grp,
 {
 	struct m0_be_log_discard_item *ldi = ast->sa_datum;
 	struct m0_be_log_discard      *ld  = ldi->ldi_ld;
+	struct m0_be_log_discard_item tmp_ldi;
+	struct m0_ext *ext;
 
-	M0_ENTRY("ld=%p ldi=%p", ld, ldi);
-
-	ld->lds_cfg.ldsc_discard(ld, ldi);
-
+	/* if lds_sync_item is NULL all the ldi are already discarded
+	   and lds_start_q will not have any ldi with LDI_FINISHED state */
+	if (ld->lds_sync_item == NULL)
+		return;
 	be_log_discard_lock(ld);
-	ldi->ldi_state = LDI_DISCARDED;
-	m0_be_log_discard_item_put(ld, ldi);
+	M0_ENTRY("ld=%p ldi=%p", ld, ldi);
+	/* capture the extent and data here in temporary ldi */
+	tmp_ldi.ldi_ext = ldi->ldi_ext;
+	tmp_ldi.ldi_user_data = m0_be_log_discard_item_user_data(ldi);
+	/* All the log discard items in ld->lds_start_q with the LDI_FINISHED state
+	   which are synced will be discarded */
+	m0_tl_for(ld_start, &ld->lds_start_q, ldi) {
+		M0_ASSERT_INFO(ldi->ldi_state == LDI_FINISHED,
+					"ldi=%p state=%d", ldi, ldi->ldi_state);
+		ld_start_tlist_del(ldi);
+		ext = m0_be_log_discard_item_ext(ldi);
+		/* update the extent with the lowest e_start and highest e_end
+		   to get the entire size of the discarded logs */
+		if (tmp_ldi.ldi_ext.e_start > ext->e_start)
+			tmp_ldi.ldi_ext.e_start = ext->e_start;
+		if (tmp_ldi.ldi_ext.e_end  < ext->e_end)
+			tmp_ldi.ldi_ext.e_end = ext->e_end;
+		if (ldi == ld->lds_sync_item)
+		{
+			ldi->ldi_state = LDI_DISCARDED;
+			m0_be_log_discard_item_put(ld, ldi);
+			break;
+		}
+		ldi->ldi_state = LDI_DISCARDED;
+		m0_be_log_discard_item_put(ld, ldi);
+	} m0_tl_endfor;
+	ld->lds_sync_item = NULL;
+	ld->lds_sync_in_progress = false;
+	if (ld_start_tlist_is_empty(&ld->lds_start_q))
+		ld->lds_sync_deadline = M0_TIME_NEVER;
+	be_log_discard_check_sync(ld, false);
+	/* The log pointers (lg_discarded position) for all the discarded items will be updated only once.
+	   This will ensure that interim lg_discarded is not picked up in lgr_last_discarded
+	   during m0_be_log_record_io_prepare() */
+	ld->lds_cfg.ldsc_discard(ld, &tmp_ldi);
```

Comment:
Please note: the `tmp_idl` only has its lid_ext and ldi_user_data valid. Others field members are not initialized. Be sure they are not used in this function.